### ISSUE TITLE
live kernel patching using kexec

### ISSUE THREAD

**Issue by copumpkin** [NixOS/nixpkgs#10726]
Could `nixos-rebuild switch` support live kernel updates via `kexec`? Should it?

**Comment by copumpkin** [comment]
Probably related to #2079 but for internal use

**Comment by CMCDragonkai** [comment]
How do you propose this to work?

**Comment by copumpkin** [comment]
Ksplice is available for free but appears unmaintained. Other than that, perhaps [KernelCare](https://en.wikipedia.org/wiki/KernelCare), [kpatch](https://github.com/dynup/kpatch), or [kGraft](https://en.wikipedia.org/wiki/KGraft) could do the trick.

**Comment by Mathnerd314** [comment]
There is [KUP](https://sslab.gtisc.gatech.edu/pages/kup.html) (still [unreleased](https://lists.openvz.org/pipermail/criu/2015-September/022035.html)); it uses CRIU/kexec and in general seems to be the most flexible approach (unlike the others it handles kernel data structure changes transparently)

**Comment by sanidhya** [comment]
I have worked on KUP prototype sometime back. It's a research prototype which works fine for most of the cases, except these: https://criu.org/What_cannot_be_checkpointed. I will be releasing the basic code by the end of January and will try to maintain it as a package. If you are interested to learn more about the complete design, then I can forward the draft privately.

**Comment by CMCDragonkai** [comment]
I'm interested to learn more about the design.

**Comment by Profpatsch** [comment]
(triage) @sanidhya did you follow up on the packaging?

**Comment by primeos** [comment]
(triage, again :)

The CRIU package is [available](https://github.com/NixOS/nixpkgs/blob/master/pkgs/os-specific/linux/criu/default.nix) by now :smile:

However, it seems like the KUP project isn't making any progress (source code is still not available and probably never will be).

Is this issue still relevant?
Imho live-patching for the kernel would be great, but we need someone to spend time on this (issue).

Some notes:
- AFAIK live-patching only works for small security-patches -> upgrading the kernel won't be possible that way
- livepatch seems like the way to go
  - Merged into the Linux kernel
  - But: only limited features yet
- If we need more features we could use [kpatch](https://github.com/dynup/kpatch) (or kGraft)

**Comment by copumpkin** [comment]
Yeah, 👍 for `livepatch`, if only because it's "vanilla" kernel

**Comment by Pitometsu** [comment]
still important to me

**Comment by Mathnerd314** [comment]
I looked at KUP again and while the source code still doesn't seem to be available, I did find the [paper](https://sanidhya.github.io/pubs/2016/kup.pdf) and [talk audio/slides](https://www.usenix.org/conference/atc16/technical-sessions/presentation/kashyap). Also there's a [Red Hat post](https://www.redhat.com/en/blog/using-criu-upgrade-vpn-servers-kernel-without-dropping-connections) which shows how to preserve one application (but not the whole system like KUP does).

The reason I'm so KUP-focused is that KUP can do updates between arbitrary versions (e.g. major updates), while the other solutions are hacky and only support minor patches. Since NixOS is based on configuration switching there's little control over the size of the kernel delta.

For this issue to move forward with KUP I see two main actions:
1. bug @sanidhya until he releases the KUP code / tells us where to find it (email on https://sanidhya.github.io/)
2. play around based on the Red Hat example with vanilla CRIU to get something like KUP but without the memory dump/load optimization

**Comment by lovesegfault** [comment]
Still important to me

**Comment by Mathnerd314** [comment]
Looking at this once more, I think although KUP's idea is sound the implementation is not worth emulating. The paper mentions that KUP requires "injecting code into the kernel binary image". But doing vanilla CRIU, there is now a page on [live migration](https://criu.org/Live_migration). Doing a live update should be very similar, replace copying with [kexec](https://wiki.archlinux.org/title/Kexec)[(8)](https://man7.org/linux/man-pages/man8/kexec.8.html), something like:
```bash
criu dump --tree 1 --images-dir /var/live-update-tmp --leave-stopped
kexec -f /efi/nixos/<hash>-linux-5.10.44-bzImage.efi --initrd=/efi/nixos/<hash>-initrd-linux-5.10.44-initrd.efi --command-line="init=/nix/store/<hash>-nixos-system-nixos-<id>/init -- restore-images-dir=/var/live-update-tmp"
```
We deliberately don't run any systemd shutdown stuff, because this is an update, not a reboot.
Then the init script sees the `restore-images-dir` parameter and runs:
```bash
criu restore --images-dir /var/live-update-tmp
```
Then assuming all went well, we're running the new kernel with the old processes. All we need to do next is activate the new configuration, like running `nixos-rebuild switch`.

So now the next step is to patch the init scripts and boot up a Linux VM to see if this approach actually works. The two failure modes I can think of are that CRIU gets stuck somewhere, or that it's too slow to be usable.

**Comment by Thesola10** [comment]
CRIU doesn't appear to want to dump itself (understandable), so it would have to be started in a `systemd` slice, dumping all other `systemd` children. What remains to be figured out is how to slide a new `systemd` underneath our newly restored process tree. Well, that or we wrap the entire running system in a thin namespace and have CRIU alongside it.

**Comment by danwdart** [comment]
Livepatch would be a good thing to revisit for a systemd service.

**Comment by danwdart** [comment]
Err, sorry, I meant the service for telling the kernel to livepatch when it needs to - rather than something to restart itself.

**Comment by Sporesirius** [comment]
Hello,
is this feature still under consideration?
It would be really nice to have rebootless kernel updates.

**Comment by Atemu** [comment]
Feel free to shoot a PR.

**Comment by tgunnoe** [comment]
semi-relevant magic
https://github.com/cleverca22/nix-tests/tree/master/kexec#readme

**Comment by Thesola10** [comment]
systemd/systemd#258 having been closed this early is not very nice, because no matter how I look at the problem, it's got to be a `systemd` issue, not a NixOS issue. After all, the easiest way to reconcile running services with `systemd`'s internal view of them... is for `systemd` itself to manage reloading.

The situation does however look suspiciously similar to `systemctl daemon-reload`, where `systemd` has to restart itself underneath a running system. I wonder if it is possible to modify this operation to dump and restore `systemd`'s understanding of the system after restoring the process tree using CRI-U. As long as CRI-U is capable of leaving restored processes in a stopped (frozen) state, we can then invoke `systemd` and have a unit fire `CONT` at everything else once our new init is ready. Remains to see how `systemd` might react to literally everything else being stopped.

**Comment by makuru-org** [comment]
I did a kexec with systemctl kexec and it worked.

**Comment by GraysonTinker** [comment]
> This is still going to require some work on the systemd+CRIU side

Systemd v261 now supports LUO, so it seems like at least some of that work has been done.
